.clang-format

```diff
@@ -0,0 +1,148 @@
+---
+Language:        Cpp
+# BasedOnStyle:  Google
```

Comment:
Can we just use this? Or clang-format does not support this by default?

Reply:
FWIW, yeti/yeti is using this:

Language: Cpp
BasedOnStyle: Google
DerivePointerAlignment: false
PointerAlignment: Left
SortIncludes: true

Reply:
And this is Android (system/core):

BasedOnStyle: Google
Standard: Cpp11
AllowShortFunctionsOnASingleLine: Inline
ColumnLimit: 100
CommentPragmas: NOLINT:.*
DerivePointerAlignment: false
IncludeBlocks: Preserve
IndentWidth: 2
PointerAlignment: Left
TabWidth: 2
UseTab: Never

Reply:
Hi both of you,
there is very little documentation out there about how clang-format handles consistency.

As far as I understood, it works the following way: (but maybe somebody of you has more insights or knows somebody, who has?)

clang-format offers backwards-compatibility. That means in this case you have a repository with a .clang-format-file generated with version X, then every version Y >= X will produce the same result on the same config file. If this were not the case, we would have to reformat our codebase for every new version/feature in clang-format. I believe that's why a "clang-format --style=Google --dump-config" produces this verbose config file, where every feature is explicitly set to a value, even if its the default value. A newer version will not enable any feature, which is not explicitly stated here. An older version, which might find a unknown config option will refuse to do its work, because this unknown option might result in different formatting introduced by a newer version of clang-format.

That said, let's consider the different options. This version:
```
---
Language: Cpp
BasedOnStyle: Google
***
```
totally works, but I couldn't figure out, whether the Google style is stable across different versions of clang-format. The oldest version I tried was 7.0.0, and the latest was something from master (not more than some weeks old). All these versions (4 in total) styled the codebase exactly the same. But I couldn't figure out whether it is guaranteed that the Google code style was the same in all its history and stays the same forever in the future. I would assume not.

The same potential problem applies to the other two examples. They are based on the "current" Google style profile and thus are tied to the "current" version of clang-format. Like I said, if the Google style never ever changes, this will be fine.

In the end, I came to the conclusion that probably the best way forward, is to stick with the recommended way (recommended by the documentation) on how to generate a .clang-format file. So that's what I've done. Taking one of the clang-format-files mentioned above might increase compatibility with older versions of clang-format, at the expense of better code styling.

Any opinions on that? Is there a minimum version we have to support? clang-7 was the oldest version available by default on my machine and it's also in Debian stable. That's why I thought it should be fine.

For reference: Here is the documentation: https://clang.llvm.org/docs/ClangFormatStyleOptions.html

Reply:
Looking at google3 and github examples, it seems to be very common to simply use something like: 

"Language: Cpp
BasedOnStyle: Google"

See abseil on github for example: https://github.com/abseil/abseil-cpp/blob/master/.clang-format.

The setting that is most commonly modified seems to be "DerivePointerAlignment".  It will "analyze the formatted file for the most common alignment of & and *".  So based on the surrounding code you'd have a different output.  I understand why many turn it off...

I'm very much in favor of using the simplest .clang-format file possible.  Even the Yeti .clang-format seems like it has some redundancies as "PointerAlignment: Left" and "SortIncludes: true" are the default values.

I'd suggest using:

Language: Cpp
BasedOnStyle: Google
DerivePointerAlignment: false

It has the added advantage of highlighting only what is different from the original style.

Also, having the simplest clang-format file will keep us in sync with any change in the default "Google" style although I doubt it will or even *can* happen.  Finally, I wouldn't worry too much about a contributor using an old version of clang-format.  It would be obvious something was wrong if a simple change touched on every single file of the project...